ipa-core/src/protocol/ipa_prf/shuffle/sharded.rs

```diff
@@ -0,0 +1,522 @@
+        RecordId,
+    },
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare, ReplicatedSecretSharing},
+        Sendable, SharedValue,
+    },
+    seq_join::{assert_send, seq_join},
+};
+
+/// This context is only useful for sharded shuffle modules because it implements common operations
+/// that all shards on all helpers perform to achieve the perfect shuffle.
+trait ShuffleContext: ShardedContext {
+    /// This sends a single machine word (8 byte value) to one of the helpers specified in
+    /// `direction` parameter.
+    fn send_word(
+        self,
+        direction: Direction,
+        val: usize,
+    ) -> impl Future<Output = Result<(), crate::error::Error>> + Send {
+        async move {
+            Ok(self
+                .set_total_records(1)
+                .send_channel::<BA64>(self.role().peer(direction))
+                .send(
+                    RecordId::FIRST,
+                    BA64::truncate_from(u128::try_from(val).unwrap()),
+                )
+                .await?)
+        }
+    }
+
+    /// This receives a single machine word (8 byte value) to one of the helpers specified in
+    /// `direction` parameter.
+    fn recv_word(
+        self,
+        direction: Direction,
+    ) -> impl Future<Output = Result<usize, crate::error::Error>> + Send {
+        async move {
+            let val: [u8; 8] = self
+                .recv_channel::<BA64>(self.role().peer(direction))
+                .receive(RecordId::FIRST)
+                .await?
+                .as_raw_slice()
+                .try_into()
+                .unwrap();
+            Ok(usize::try_from(u64::from_le_bytes(val)).unwrap())
+        }
+    }
+
+    /// In sharded shuffle, it is a common operation to apply a mask to the input and then permute
+    /// it, with all shards involved in the process. This routine implements it, so helper-specific
+    /// code does not need to do this repetitive work.
+    ///
+    /// The destination shard for each masked row is decided based on value obtained from sampling
+    /// PRSS. Which value to use (left or right) is decided based on `direction` parameter.
+    fn mask_and_shuffle<I, S>(
+        self,
+        direction: Direction,
+        data: I,
+    ) -> impl Future<Output = Result<Vec<S>, crate::error::Error>> + Send
+    where
+        I: IntoIterator<Item = S>,
+        I::IntoIter: ExactSizeIterator + Send,
+        S: ShuffleShare,
+    {
+        let data = data.into_iter();
+        async move {
+            let masking_ctx = self.narrow(&ShuffleStep::Mask);
+            let mut resharded = assert_send(reshard(
+                self.clone(),
+                data.enumerate().map(|(i, item)| {
+                    let (l, r) = masking_ctx.prss().generate(RecordId::from(i));
+                    let mask = match direction {
+                        Direction::Left => l,
+                        Direction::Right => r,
```

Comment:
shouldnt it be that if I send the mask to the left, I need to generate it via the prss part on the right? Since the left party can also generate the same mask which would violate security.

Reply:
> is it ok to throw away the prss part that you don't use? You would probably need to generate it again. Wouldn't prss complain about double use?

it would complain about double use, I use grouping steps like `Permute12`, `Permute13` to get a unique AES key per masking step


> shouldnt it be that if I send the mask to the left, I need to generate it via the prss part on the right? Since the left party can also generate the same mask which would violate security.

Correct, before revealing private data to the left, it must be masked with values obtained from PRSS shared with the right helper. This function however does not reveal data and does not make use of MPC channels.